File: dt241m-controller/src/mqtt/discovery.ts
import { MAX_NAME_LENGTH, displayName, type Adapter } from "../registry.js";
import {
  CONTROLLER_DEVICE_IDENTIFIER,
  CONTROLLER_NODE_ID,
  PAYLOAD_OFFLINE,
  PAYLOAD_ONLINE,
  PAYLOAD_PRESS,
  controllerTopics,
  deviceIdentifier,
  deviceNodeId,
  deviceTopics,
  haDiscoveryTopic
} from "./topics.js";

export type DiscoveryMessage = { topic: string; payload: Record<string, unknown> };

export type DiscoveryContext = {
  version: string;
  supportUrl: string;
};

const MANUFACTURER = "PWAY";
const CONTROLLER_NAME = "DT241M Controller";

export const ICONS = {
  receiver: "mdi:monitor",
  transmitter: "mdi:broadcast",
  unknown: "mdi:help-network",
  controller: "mdi:video-switch",
  rescan: "mdi:radar",
  name: "mdi:rename-box",
  ip: "mdi:ip-network",
  role: "mdi:swap-horizontal"
} as const;

function origin(ctx: DiscoveryContext) {
  return { name: CONTROLLER_NAME, sw_version: ctx.version, support_url: ctx.supportUrl };
}

function controllerAvailability() {
  return {
    topic: controllerTopics.availability,
    payload_available: PAYLOAD_ONLINE,
    payload_not_available: PAYLOAD_OFFLINE
  };
}

function controllerDevice(ctx: DiscoveryContext) {
  return {
    identifiers: [CONTROLLER_DEVICE_IDENTIFIER],
    name: CONTROLLER_NAME,
    manufacturer: "dt241m-controller add-on",
    model: "DT241M MQTT bridge",
    sw_version: ctx.version
  };
}

function adapterDevice(adapter: Adapter) {
  return {
    identifiers: [deviceIdentifier(adapter.mac)],
    connections: [["mac", adapter.mac]],
    name: displayName(adapter),
    manufacturer: MANUFACTURER,
    model: adapter.productName ?? adapter.model ?? "DT241M",
    model_id: adapter.model ?? undefined,
    sw_version: adapter.firmware ?? undefined,
    via_device: CONTROLLER_DEVICE_IDENTIFIER
  };
}

function adapterAvailability(adapter: Adapter) {
  const topics = deviceTopics(adapter.mac);
  return {
    availability: [
      controllerAvailability(),
      { topic: topics.availability, payload_available: PAYLOAD_ONLINE, payload_not_available: PAYLOAD_OFFLINE }
    ],
    availability_mode: "all"
  };
}

export function receiverChannelDiscovery(adapter: Adapter, ctx: DiscoveryContext): DiscoveryMessage {
  const topics = deviceTopics(adapter.mac);
  return {
    topic: haDiscoveryTopic("number", deviceNodeId(adapter.mac), "channel"),
    payload: {
      name: "Channel",
      unique_id: `${adapter.id}_channel`,
      object_id: `${adapter.id}_channel`,
      state_topic: topics.channelState,
      command_topic: topics.channelSet,
      min: 0,
      max: 255,
      step: 1,
      mode: "box",
      optimistic: false,
      retain: false,
      qos: 1,
      icon: ICONS.receiver,
      ...adapterAvailability(adapter),
      device: adapterDevice(adapter),
      origin: origin(ctx)
    }
  };
}

export function readOnlyChannelDiscovery(adapter: Adapter, ctx: DiscoveryContext): DiscoveryMessage {
  const topics = deviceTopics(adapter.mac);
  return {
    topic: haDiscoveryTopic("sensor", deviceNodeId(adapter.mac), "channel"),
    payload: {
      name: "Channel",
      unique_id: `${adapter.id}_channel`,
      object_id: `${adapter.id}_channel`,
      state_topic: topics.channelState,
      icon: ICONS[adapter.role],
      ...adapterAvailability(adapter),
      device: adapterDevice(adapter),
      origin: origin(ctx)
    }
  };
}

export function ipAddressDiscovery(adapter: Adapter, ctx: DiscoveryContext): DiscoveryMessage {
  const topics = deviceTopics(adapter.mac);
  return {
    topic: haDiscoveryTopic("sensor", deviceNodeId(adapter.mac), "ip_address"),
    payload: {
      name: "IP address",
      unique_id: `${adapter.id}_ip_address`,
      object_id: `${adapter.id}_ip_address`,
      state_topic: topics.ipState,
      entity_category: "diagnostic",
      icon: ICONS.ip,
      ...adapterAvailability(adapter),
      device: adapterDevice(adapter),
      origin: origin(ctx)
    }
  };
}

export function nameDiscovery(adapter: Adapter, ctx: DiscoveryContext): DiscoveryMessage {
  const topics = deviceTopics(adapter.mac);
  return {
    topic: haDiscoveryTopic("text", deviceNodeId(adapter.mac), "name"),
    payload: {
      name: "Name",
      unique_id: `${adapter.id}_name`,
      object_id: `${adapter.id}_name`,
      state_topic: topics.nameState,
      command_topic: topics.nameSet,
      min: 0,
      max: MAX_NAME_LENGTH,
      mode: "text",
      retain: false,
      qos: 1,
      entity_category: "config",
      icon: ICONS.name,
      availability: [controllerAvailability()],
      device: adapterDevice(adapter),
      origin: origin(ctx)
    }
  };
}

export function roleDiscovery(adapter: Adapter, ctx: DiscoveryContext): DiscoveryMessage {
  const topics = deviceTopics(adapter.mac);
  return {
    topic: haDiscoveryTopic("sensor", deviceNodeId(adapter.mac), "role"),
    payload: {
      name: "Role",
      unique_id: `${adapter.id}_role`,
      object_id: `${adapter.id}_role`,
      state_topic: topics.roleState,
      entity_category: "diagnostic",
      icon: ICONS.role,
      ...adapterAvailability(adapter),
      device: adapterDevice(adapter),
      origin: origin(ctx)
    }
  };
}

export function adapterDiscoveryMessages(adapter: Adapter, ctx: DiscoveryContext): DiscoveryMessage[] {
  const channel = adapter.role === "receiver" ? receiverChannelDiscovery(adapter, ctx) : readOnlyChannelDiscovery(adapter, ctx);
  return [channel, nameDiscovery(adapter, ctx), ipAddressDiscovery(adapter, ctx), roleDiscovery(adapter, ctx)];
}

export function staleAdapterDiscoveryTopics(adapter: Adapter): string[] {
  const other = adapter.role === "receiver" ? "sensor" : "number";
  return [haDiscoveryTopic(other, deviceNodeId(adapter.mac), "channel")];
}

export function rescanButtonDiscovery(ctx: DiscoveryContext): DiscoveryMessage {
  return {
    topic: haDiscoveryTopic("button", CONTROLLER_NODE_ID, "rescan"),
    payload: {
      name: "Rescan network",
      unique_id: `${CONTROLLER_NODE_ID}_rescan`,
      object_id: `${CONTROLLER_NODE_ID}_rescan`,
      command_topic: controllerTopics.rescanPress,
      payload_press: PAYLOAD_PRESS,
      retain: false,
      qos: 1,
      icon: ICONS.rescan,
      availability: [controllerAvailability()],
      device: controllerDevice(ctx),
      origin: origin(ctx)
    }
  };
}

export function knownDevicesDiscovery(ctx: DiscoveryContext): DiscoveryMessage {
  return {
    topic: haDiscoveryTopic("sensor", CONTROLLER_NODE_ID, "known_devices"),
    payload: {
      name: "Known devices",
      unique_id: `${CONTROLLER_NODE_ID}_known_devices`,
      object_id: `${CONTROLLER_NODE_ID}_known_devices`,
      state_topic: controllerTopics.knownDevicesState,
      state_class: "measurement",
      icon: ICONS.controller,
      availability: [controllerAvailability()],
      device: controllerDevice(ctx),
      origin: origin(ctx)
    }
  };
}

export function onlineDevicesDiscovery(ctx: DiscoveryContext): DiscoveryMessage {
  return {
    topic: haDiscoveryTopic("sensor", CONTROLLER_NODE_ID, "online_devices"),
    payload: {
      name: "Online devices",
      unique_id: `${CONTROLLER_NODE_ID}_online_devices`,
      object_id: `${CONTROLLER_NODE_ID}_online_devices`,
      state_topic: controllerTopics.onlineDevicesState,
      state_class: "measurement",
      icon: ICONS.controller,
      availability: [controllerAvailability()],
      device: controllerDevice(ctx),
      origin: origin(ctx)
    }
  };
}

export function controllerDiscoveryMessages(ctx: DiscoveryContext): DiscoveryMessage[] {
  return [rescanButtonDiscovery(ctx), knownDevicesDiscovery(ctx), onlineDevicesDiscovery(ctx)];
}

File: dt241m-controller/internal/mqtt/discovery.go
package mqtt

import (
	"encoding/json"

	"github.com/jacobgad/dt241m-controller/internal/dt241m"
	"github.com/jacobgad/dt241m-controller/internal/registry"
)

// Origin identifies this add-on in discovery payloads.
type Origin struct {
	Version    string
	SupportURL string
}

// Message is one retained discovery config.
type Message struct {
	Topic   string
	Payload map[string]any
}

// JSON renders the payload.
func (m Message) JSON() string {
	data, _ := json.Marshal(m.Payload)
	return string(data)
}

const (
	manufacturer   = "PWAY"
	controllerName = "DT241M Controller"
)

const (
	iconController = "mdi:video-switch"
	iconRescan     = "mdi:radar"
	iconName       = "mdi:rename-box"
	iconIP         = "mdi:ip-network"
	iconRole       = "mdi:swap-horizontal"
	iconSource     = "mdi:video-input-hdmi"
)

func roleIcon(role dt241m.Role) string {
	switch role {
	case dt241m.RoleReceiver:
		return "mdi:monitor"
	case dt241m.RoleTransmitter:
		return "mdi:broadcast"
	default:
		return "mdi:help-network"
	}
}

func origin(o Origin) map[string]any {
	return map[string]any{"name": controllerName, "sw_version": o.Version, "support_url": o.SupportURL}
}

func controllerAvailability() map[string]any {
	return map[string]any{"topic": ControllerAvailability, "payload_available": PayloadOnline, "payload_not_available": PayloadOffline}
}

func controllerDevice(o Origin) map[string]any {
	return map[string]any{
		"identifiers":  []string{ControllerIdentifier},
		"name":         controllerName,
		"manufacturer": "dt241m-controller add-on",
		"model":        "DT241M MQTT bridge",
		"sw_version":   o.Version,
	}
}

func adapterDevice(a registry.Adapter) map[string]any {
	device := map[string]any{
		"identifiers":  []string{DeviceIdentifier(a.MAC)},
		"connections":  [][]string{{"mac", a.MAC}},
		"name":         a.DisplayName(),
		"manufacturer": manufacturer,
		"via_device":   ControllerIdentifier,
	}
	switch {
	case a.ProductName != "":
		device["model"] = a.ProductName
	case a.Model != "":
		device["model"] = a.Model
	default:
		device["model"] = "DT241M"
	}
	if a.Model != "" {
		device["model_id"] = a.Model
	}
	if a.Firmware != "" {
		device["sw_version"] = a.Firmware
	}
	return device
}

func withAdapterAvailability(a registry.Adapter, payload map[string]any) map[string]any {
	payload["availability"] = []map[string]any{
		controllerAvailability(),
		{"topic": ForDevice(a.MAC).Availability, "payload_available": PayloadOnline, "payload_not_available": PayloadOffline},
	}
	payload["availability_mode"] = "all"
	return payload
}

// ReceiverChannel is the writable Channel number entity for a receiver.
func ReceiverChannel(a registry.Adapter, o Origin) Message {
	t := ForDevice(a.MAC)
	return Message{
		Topic: HADiscoveryTopic("number", DeviceNodeID(a.MAC), "channel"),
		Payload: withAdapterAvailability(a, map[string]any{
			"name":          "Channel",
			"unique_id":     a.ID + "_channel",
			"object_id":     a.ID + "_channel",
			"state_topic":   t.ChannelState,
			"command_topic": t.ChannelSet,
			"min":           dt241m.MinChannel,
			"max":           dt241m.MaxChannel,
			"step":          1,
			"mode":          "box",
			"optimistic":    false,
			"retain":        false,
			"qos":           1,
			"icon":          roleIcon(dt241m.RoleReceiver),
			"device":        adapterDevice(a),
			"origin":        origin(o),
		}),
	}
}

// TransmitterChannel is the writable Channel number for a transmitter. It is a
// configuration entity: changing it re-routes every receiver watching that transmitter.
func TransmitterChannel(a registry.Adapter, o Origin) Message {
	t := ForDevice(a.MAC)
	return Message{
		Topic: HADiscoveryTopic("number", DeviceNodeID(a.MAC), "channel"),
		Payload: withAdapterAvailability(a, map[string]any{
			"name":            "Channel",
			"unique_id":       a.ID + "_channel",
			"object_id":       a.ID + "_channel",
			"state_topic":     t.ChannelState,
			"command_topic":   t.ChannelSet,
			"min":             dt241m.MinChannel,
			"max":             dt241m.MaxChannel,
			"step":            1,
			"mode":            "box",
			"optimistic":      false,
			"retain":          false,
			"qos":             1,
			"entity_category": "config",
			"icon":            roleIcon(dt241m.RoleTransmitter),
			"device":          adapterDevice(a),
			"origin":          origin(o),
		}),
	}
}

// ReceiverSource is the Source select on a receiver: pick a transmitter by name instead
// of remembering its channel. The state is "none" when no transmitter uses the channel.
func ReceiverSource(a registry.Adapter, options []string, o Origin) Message {
	t := ForDevice(a.MAC)
	if options == nil {
		options = []string{}
	}
	return Message{
		Topic: HADiscoveryTopic("select", DeviceNodeID(a.MAC), "source"),
		Payload: withAdapterAvailability(a, map[string]any{
			"name":          "Source",
			"unique_id":     a.ID + "_source",
			"object_id":     a.ID + "_source",
			"state_topic":   t.SourceState,
			"command_topic": t.SourceSet,
			"options":       options,
			"optimistic":    false,
			"retain":        false,
			"qos":           1,
			"icon":          iconSource,
			"device":        adapterDevice(a),
			"origin":        origin(o),
		}),
	}
}

// ReadOnlyChannel is the Channel sensor for devices whose role could not be determined.
func ReadOnlyChannel(a registry.Adapter, o Origin) Message {
	t := ForDevice(a.MAC)
	return Message{
		Topic: HADiscoveryTopic("sensor", DeviceNodeID(a.MAC), "channel"),
		Payload: withAdapterAvailability(a, map[string]any{
			"name":        "Channel",
			"unique_id":   a.ID + "_channel",
			"object_id":   a.ID + "_channel",
			"state_topic": t.ChannelState,
			"icon":        roleIcon(a.Role),
			"device":      adapterDevice(a),
			"origin":      origin(o),
		}),
	}
}

// Name is the writable Name text entity; available whenever the controller is, so offline devices can still be renamed.
func Name(a registry.Adapter, o Origin) Message {
	t := ForDevice(a.MAC)
	return Message{
		Topic: HADiscoveryTopic("text", DeviceNodeID(a.MAC), "name"),
		Payload: map[string]any{
			"name":            "Name",
			"unique_id":       a.ID + "_name",
			"object_id":       a.ID + "_name",
			"state_topic":     t.NameState,
			"command_topic":   t.NameSet,
			"min":             0,
			"max":             registry.MaxNameLength,
			"mode":            "text",
			"retain":          false,
			"qos":             1,
			"entity_category": "config",
			"icon":            iconName,
			"availability":    []map[string]any{controllerAvailability()},
			"device":          adapterDevice(a),
			"origin":          origin(o),
		},
	}
}

// IPAddress is the diagnostic IP sensor.
func IPAddress(a registry.Adapter, o Origin) Message {
	return Message{
		Topic: HADiscoveryTopic("sensor", DeviceNodeID(a.MAC), "ip_address"),
		Payload: withAdapterAvailability(a, map[string]any{
			"name":            "IP address",
			"unique_id":       a.ID + "_ip_address",
			"object_id":       a.ID + "_ip_address",
			"state_topic":     ForDevice(a.MAC).IPState,
			"entity_category": "diagnostic",
			"icon":            iconIP,
			"device":          adapterDevice(a),
			"origin":          origin(o),
		}),
	}
}

// Role is the diagnostic Role sensor.
func Role(a registry.Adapter, o Origin) Message {
	return Message{
		Topic: HADiscoveryTopic("sensor", DeviceNodeID(a.MAC), "role"),
		Payload: withAdapterAvailability(a, map[string]any{
			"name":            "Role",
			"unique_id":       a.ID + "_role",
			"object_id":       a.ID + "_role",
			"state_topic":     ForDevice(a.MAC).RoleState,
			"entity_category": "diagnostic",
			"icon":            iconRole,
			"device":          adapterDevice(a),
			"origin":          origin(o),
		}),
	}
}

// AdapterMessages lists every entity published for an adapter; sourceOptions is only used for receivers.
func AdapterMessages(a registry.Adapter, sourceOptions []string, o Origin) []Message {
	messages := []Message{Name(a, o), IPAddress(a, o), Role(a, o)}
	switch a.Role {
	case dt241m.RoleReceiver:
		messages = append(messages, ReceiverChannel(a, o), ReceiverSource(a, sourceOptions, o))
	case dt241m.RoleTransmitter:
		messages = append(messages, TransmitterChannel(a, o))
	default:
		messages = append(messages, ReadOnlyChannel(a, o))
	}
	return messages
}

// StaleTopics are config topics that must be cleared for an adapter in its current role,
// so that a unique_id never lingers under a platform it no longer uses (role flips, and
// the 2.0 → 2.1 move of the transmitter channel from sensor to number).
func StaleTopics(a registry.Adapter) []string {
	node := DeviceNodeID(a.MAC)
	switch a.Role {
	case dt241m.RoleReceiver:
		return []string{HADiscoveryTopic("sensor", node, "channel")}
	case dt241m.RoleTransmitter:
		return []string{HADiscoveryTopic("sensor", node, "channel"), HADiscoveryTopic("select", node, "source")}
	default:
		return []string{HADiscoveryTopic("number", node, "channel"), HADiscoveryTopic("select", node, "source")}
	}
}

// RescanButton is the controller's Rescan network button.
func RescanButton(o Origin) Message {
	return Message{
		Topic: HADiscoveryTopic("button", ControllerNodeID, "rescan"),
		Payload: map[string]any{
			"name":          "Rescan network",
			"unique_id":     ControllerNodeID + "_rescan",
			"object_id":     ControllerNodeID + "_rescan",
			"command_topic": ControllerRescanPress,
			"payload_press": PayloadPress,
			"retain":        false,
			"qos":           1,
			"icon":          iconRescan,
			"availability":  []map[string]any{controllerAvailability()},
			"device":        controllerDevice(o),
			"origin":        origin(o),
		},
	}
}

func countSensor(o Origin, objectID, name, stateTopic string) Message {
	return Message{
		Topic: HADiscoveryTopic("sensor", ControllerNodeID, objectID),
		Payload: map[string]any{
			"name":         name,
			"unique_id":    ControllerNodeID + "_" + objectID,
			"object_id":    ControllerNodeID + "_" + objectID,
			"state_topic":  stateTopic,
			"state_class":  "measurement",
			"icon":         iconController,
			"availability": []map[string]any{controllerAvailability()},
			"device":       controllerDevice(o),
			"origin":       origin(o),
		},
	}
}

// ControllerMessages lists the controller device's entities.
func ControllerMessages(o Origin) []Message {
	return []Message{
		RescanButton(o),
		countSensor(o, "known_devices", "Known devices", ControllerKnownState),
		countSensor(o, "online_devices", "Online devices", ControllerOnlineState),
	}
}
